storage/index/results.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package index
+
+import (
+	"errors"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3x/ident"
+	"github.com/m3db/m3x/pool"
+)
+
+var (
+	errUnableToAddDocMissingID = errors.New("corrupt data, unable to extract id")
+)
+
+type results struct {
+	nsID       ident.ID
+	size       int
+	resultsMap *ResultsMap
+
+	idPool       ident.Pool
+	bytesPool    pool.CheckedBytesPool
+	tagArrayPool TagArrayPool
+
+	pool ResultsPool
+}
+
+// NewResults returns a new results object.
+func NewResults(opts Options) Results {
+	return &results{
+		resultsMap:   newResultsMap(opts.IdentifierPool()),
+		idPool:       opts.IdentifierPool(),
+		bytesPool:    opts.CheckedBytesPool(),
+		tagArrayPool: opts.TagArrayPool(),
+		pool:         opts.ResultsPool(),
+	}
+}
+
+func (r *results) Add(d doc.Document) (added bool, size int, err error) {
+	added = false
+	if len(d.ID) == 0 {
+		return added, r.size, errUnableToAddDocMissingID
+	}
+
+	// NB: can cast the []byte -> ident.ID to avoid an alloc
+	// before we're sure we need it.
+	tsID := ident.BytesID(d.ID)
+
+	// check if it already exists in the map.
+	if r.resultsMap.Contains(tsID) {
+		return added, r.size, nil
+	}
+
+	// i.e. it doesn't exist in the map, so we create the tags wrapping
+	// fields prodided by the document.
+	tags := r.tags(d.Fields)
+
+	// We use Set() instead of SetUnsafe to ensure we're taking a copy of
+	// the tsID's bytes.
+	r.resultsMap.Set(tsID, tags)
+	r.size++
+
+	added = true
+	return added, r.size, nil
+}
+
+func (r *results) tags(fields doc.Fields) []ident.Tag {
+	tags := r.tagArrayPool.Get()[:0]
```

Comment:
Hm should we perhaps make the pool do the `[:0]` subslice either on the get or the put?

Reply:
Sure, the pool already does the `[:0]` on Put, will remove here.

Reply:
done